src/main/main.c

```diff
@@ -646,33 +736,45 @@ static int fd_poll(struct re *re)
 #ifdef HAVE_SELECT
 	case METHOD_SELECT: {
 		struct timeval tv;
+		int max_fd = 0;
+		int cfds = 0;
 
 		/* Clear and update fd sets */
 		FD_ZERO(&rfds);
 		FD_ZERO(&wfds);
 		FD_ZERO(&efds);
 
-		for (i=0; i<re->nfds; i++) {
-			re_sock_t fd = re->fhs[i].fd;
-			if (!re->fhs[i].fh)
+		for (i = 0; cfds < nfds; i++) {
+			fhs = re->fhsl[i];
+
+			if (!fhs || !fhs->fh)
 				continue;
 
-			if (re->fhs[i].flags & FD_READ)
+			++cfds;
+
+			re_sock_t fd = fhs->fd;
+			if (fhs->flags & FD_READ)
 				FD_SET(fd, &rfds);
-			if (re->fhs[i].flags & FD_WRITE)
+			if (fhs->flags & FD_WRITE)
 				FD_SET(fd, &wfds);
-			if (re->fhs[i].flags & FD_EXCEPT)
+			if (fhs->flags & FD_EXCEPT)
 				FD_SET(fd, &efds);
+#if !defined(WIN32)
+			max_fd = max(max_fd, fd + 1);
+#endif
 		}
 
 #ifdef WIN32
 		tv.tv_sec  = (long) to / 1000;
+		nfds = i;
 #else
 		tv.tv_sec  = (time_t) to / 1000;
+		nfds = max_fd;
```

Comment:
Why? (also in the `WIN32` branch above)

`nfds` is the number of elements in the `fhsl` list, not the max fd number + 1. In the loop below, you are using `nfds` to limit `i`, where `i` is used to index into `fhsl`.